NServiceBus.FileBasedRouting/PublishRoutingConnector.cs

```diff
@@ -0,0 +1,103 @@
+using System.Linq;
+using System.Threading.Tasks;
+using NServiceBus.Pipeline;
+using NServiceBus.Routing;
+using NServiceBus.Unicast.Queuing;
+
+namespace NServiceBus.FileBasedRouting
+{
+    class PublishRoutingConnector : StageConnector<IOutgoingPublishContext, IOutgoingLogicalMessageContext>
+    {
+        readonly UnicastSubscriberTable routingTable;
+        readonly EndpointInstances endpointInstances;
+        readonly IDistributionPolicy distributionPolicy;
+        readonly Func<EndpointInstance, string> resolveTransportAddress;
+
+        public PublishRoutingConnector(UnicastSubscriberTable routingTable, EndpointInstances endpointInstances, IDistributionPolicy distributionPolicy, Func<EndpointInstance, string> resolveTransportAddress)
+        {
+            this.routingTable = routingTable;
+            this.endpointInstances = endpointInstances;
+            this.distributionPolicy = distributionPolicy;
+            this.resolveTransportAddress = resolveTransportAddress;
+        }
+
+        public override async Task Invoke(IOutgoingPublishContext context, Func<IOutgoingLogicalMessageContext, Task> stage)
+        {
+            var eventType = context.Message.MessageType;
+            var routingStrategies = Route(eventType).ToList();
+            if (routingStrategies.Count == 0)
+            {
+                //No subscribers for this message.
+                return;
+            }
+
+            context.Headers[Headers.MessageIntent] = MessageIntentEnum.Publish.ToString();
+
+            try
+            {
+                await stage(this.CreateOutgoingLogicalMessageContext(context.Message, routingStrategies, context)).ConfigureAwait(false);
+            }
+            catch (QueueNotFoundException ex)
+            {
+                throw new Exception($"The destination queue '{ex.Queue}' could not be found. The destination may be misconfigured for this kind of message ({eventType}) in the MessageEndpointMappings of the UnicastBusConfig section in the configuration file. It may also be the case that the given queue hasn\'t been created yet, or has been deleted.", ex);
+            }
+        }
+
+        IEnumerable<UnicastRoutingStrategy> Route(Type messageType)
+        {
+            var routes = routingTable.GetRoutesFor(messageType);
+            var selectedDestinations = SelectDestinationsForEachEndpoint(routes);
+            return selectedDestinations.Select(destination => new UnicastRoutingStrategy(destination));
+        }
+
+        HashSet<string> SelectDestinationsForEachEndpoint(IEnumerable<UnicastRouteGroup> routeGroups)
+        {
+            //Make sure we are sending only one to each transport destination. Might happen when there are multiple routing information sources.
+            var addresses = new HashSet<string>();
+
+            foreach (var group in routeGroups)
+            {
+                if (group.EndpointName == null) //Routing targets that do not specify endpoint name
```

Comment:
In the current implementation -> never. That's why I left a comment about possible simplification. As long we keep (re)using `UnicastRoute` I'd vote too keep this branch because `UnciastRoute` allows for it. I was also thinking about extending the file format in order to allow for putting transport address instead of endpoint logical name in the XML. The this branch would make sense.